### ISSUE TITLE
Problems with Pocketbook Touch Lux 5

### ISSUE THREAD

**Issue by Engberg-Jacob** [zisoft/calibre-pocketbook-device-driver#3]
Hello!

I have a Pocketbook Touch Lux 5, and am using Calibre 5.35, Win65. When I do the steps described in the ReadMe, I can't get the reading status to sync.

I have tried using the Debug Mode, and the only lines that might be relevant I could see was:

```
DEBUG:    0.0 POCKETBOOK632: open()
DEBUG:    0.0 POCKETBOOK632: read lookup name: #readornot
DEBUG:    0.0 POCKETBOOK632: Found database at path I:\system\explorer-3\explorer-3.db
DEBUG:    0.0 POCKETBOOK632: database cleanup
DEBUG:    0.0 POCKETBOOK632: 0 rows from books_settings deleted
DEBUG:    0.0 POCKETBOOK632: 0 rows from booktogenre deleted
DEBUG:    0.0 POCKETBOOK632: 0 rows from social deleted
DEBUG:    0.1 POCKETBOOK632: 0 rows from books_impl deleted
DEBUG:    0.1 POCKETBOOK632: 0 rows from books_fast_hashes deleted
```
and
```
DEBUG:    0.2 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= None
DEBUG:    0.2 USBMS: dirs are: I:\ Books
DEBUG:    0.3 USBMS: scan from root True I:\
DEBUG:    0.4 USBMS: count found in cache: 130, count of files in metadata: 130, need_sync: False
DEBUG:    0.4 USBMS: Finished fetching list of books from device. oncard= None
DEBUG:    0.4 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= carda
DEBUG:    0.4 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= cardb
```
I don't know if it is my eReader, and therefor outside the scope of this plugin, but I thought I would draw your attention to it.

**Comment by zisoft** [comment]
It seems that the standard USBMS driver is used here. Can you please check the log for all lines containing "POCKETBOOK632" and post here.

There should at least be a line saying 'POCKETBOOK632: start sychronize_with_db'

**Comment by zisoft** [comment]
Can you please go to Calibre --> Preferences --> Advanced --> Miscellaneous --> Debug device detection
and follow the instructions there.
Then post the Device information here.

For my PB632 it is:

USB Vendor ID (in hex): 0xfffe
USB Product ID (in hex): 0x1
USB Revision ID (in hex): 0x101

**Comment by Engberg-Jacob** [comment]
I couldn't finde the line you referenced here,
> There should at least be a line saying 'POCKETBOOK632: start sychronize_with_db'

I was only able to find
```
Successfully initialized third party plugins: DeDRM (7, 2, 1) && EpubSplit (3, 3, 0) && KOReader Sync (0, 2, 7) && KePub Metadata Writer (3, 4, 3) && PocketBook632 (1, 0, 2)
```
```
DEBUG:    5.4 POCKETBOOK632: open()
DEBUG:    5.4 POCKETBOOK632: read lookup name: #readornot
DEBUG:    5.4 POCKETBOOK632: Found database at path I:\system\explorer-3\explorer-3.db
DEBUG:    5.4 POCKETBOOK632: database cleanup
DEBUG:    5.5 POCKETBOOK632: 0 rows from books_settings deleted
DEBUG:    5.5 POCKETBOOK632: 0 rows from booktogenre deleted
DEBUG:    5.5 POCKETBOOK632: 0 rows from social deleted
DEBUG:    5.5 POCKETBOOK632: 0 rows from books_impl deleted
DEBUG:    5.5 POCKETBOOK632: 0 rows from books_fast_hashes deleted
```
```
DEBUG:    5.8 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= None
```
```
DEBUG:    6.4 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= carda
DEBUG:    6.4 USBMS: Fetching list of books from device. Device= POCKETBOOK632 oncard= cardb
```

I also tried to debug the device detection, but when my device is plugged in, it doesn't result in anything, even after an hour.
If no device is plugged in, it results in all these device connections
```
(USBDevice(vendor_id=None product_id=None bcd=None devid=usb\root_hub30&vid1022&pid149c&rev0000 devinst=4),
 USBDevice(vendor_id=0x46d product_id=0xc531 bcd=0x2100 devid=usb\vid_046d&pid_c531&rev_2100 devinst=5),
 USBDevice(vendor_id=None product_id=None bcd=None devid=usb\root_hub30&vid1022&pid149c&rev0000 devinst=6),
 USBDevice(vendor_id=0x1462 product_id=0x7b93 bcd=0x1 devid=usb\vid_1462&pid_7b93&rev_0001 devinst=7),
 USBDevice(vendor_id=0x46d product_id=0xc531 bcd=0x2100 devid=usb\vid_046d&pid_c531&rev_2100&mi_01 devinst=8),
 USBDevice(vendor_id=0xd8c product_id=0x44 bcd=0x14 devid=usb\vid_0d8c&pid_0044&rev_0014 devinst=9),
 USBDevice(vendor_id=0xd8c product_id=0x44 bcd=0x14 devid=usb\vid_0d8c&pid_0044&rev_0014&mi_03 devinst=10),
 USBDevice(vendor_id=None product_id=None bcd=None devid=usb\root_hub30&vid10de&pid1ad8&rev00a1 devinst=11),
 USBDevice(vendor_id=0x46d product_id=0xc07c bcd=0x8500 devid=usb\vid_046d&pid_c07c&rev_8500&mi_00 devinst=12),
 USBDevice(vendor_id=0x46d product_id=0xc07c bcd=0x8500 devid=usb\vid_046d&pid_c07c&rev_8500 devinst=13),
 USBDevice(vendor_id=None product_id=None bcd=None devid=usb\root_hub30&vid1022&pid145f&rev0000 devinst=14),
 USBDevice(vendor_id=0x46d product_id=0x85c bcd=0x16 devid=usb\vid_046d&pid_085c&rev_0016 devinst=15),
 USBDevice(vendor_id=0x46d product_id=0xc07c bcd=0x8500 devid=usb\vid_046d&pid_c07c&rev_8500&mi_01 devinst=16),
 USBDevice(vendor_id=0x46d product_id=0xc24d bcd=0x8001 devid=usb\vid_046d&pid_c24d&rev_8001&mi_00 devinst=17),
 USBDevice(vendor_id=0x46d product_id=0xc24d bcd=0x8001 devid=usb\vid_046d&pid_c24d&rev_8001 devinst=18),
 USBDevice(vendor_id=0x5e3 product_id=0x608 bcd=0x8832 devid=usb\vid_05e3&pid_0608&rev_8832 devinst=19),
 USBDevice(vendor_id=0x46d product_id=0xc531 bcd=0x2100 devid=usb\vid_046d&pid_c531&rev_2100&mi_00 devinst=20),
 USBDevice(vendor_id=0xd8c product_id=0x44 bcd=0x14 devid=usb\vid_0d8c&pid_0044&rev_0014&mi_00 devinst=21),
 USBDevice(vendor_id=0x8087 product_id=0x29 bcd=0x1 devid=usb\vid_8087&pid_0029&rev_0001 devinst=22),
 USBDevice(vendor_id=0x46d product_id=0x85c bcd=0x16 devid=usb\vid_046d&pid_085c&rev_0016&mi_02 devinst=23),
 USBDevice(vendor_id=0x46d product_id=0x85c bcd=0x16 devid=usb\vid_046d&pid_085c&rev_0016&mi_00 devinst=24),
 USBDevice(vendor_id=0x46d product_id=0xc24d bcd=0x8001 devid=usb\vid_046d&pid_c24d&rev_8001&mi_01 devinst=25))
```

**Comment by zisoft** [comment]
`DEBUG:    5.4 POCKETBOOK632: read lookup name: #readornot`

Did you create a userdefined Yes/No column with that lookup name?

**Comment by Engberg-Jacob** [comment]
Yes, I did. I thought that maybe have been an issue in the start, which is why I changed the default, to make sure it was the same.
I've attached a picture to show the userdefined column, and the plugin settings

![billede](https://user-images.githubusercontent.com/55184216/184536187-287c811f-af81-4673-b78e-2ff6a17410a3.png)

![billede](https://user-images.githubusercontent.com/55184216/184538312-ac740ae4-ccc1-43bb-a128-7d85378b705a.png)

**Comment by zisoft** [comment]
Can you please do the following steps:

- start Calibre with device DISCONNECTED
- go to preferences, advanced-->miscellaneous-->"Get information to setup the user defined device" (second button)
- ensure your device is disconnected, press ok
- now plug in your device and wait until Calibre detects it (may take up to 10-20 seconds), the press ok
- copy & paste the device values here

**Comment by Engberg-Jacob** [comment]
Thank you for the help :)

Device values:

USB Vendor ID (in hex): 0xfffe
USB Product ID (in hex): 0x1
USB Revision ID (in hex): 0x101

**Comment by zisoft** [comment]
These are the same identifiers as for my PocketBook HD Touch 3, which identifies as a PB632 device.
So it should simply work with your device, too.

Do you have any books marked as "read" on your device?

**Comment by Engberg-Jacob** [comment]
Yes, several. But none of them actually show up in the userdefined Calibre column. I have also tried marking a books a "unread", then "read", to make sure it actually is marked as "read". They aren't imported in the column either. Neither are the books that I have simply read to the last page. Is there something else that might be wrong? Am I misunderstanding the plugin? It should import the "read" status from the Pocketbook, right?

Should I take some pictures of the read status on the Pocketbook, to verify I have done the correct thing? Thank you

**Comment by CZ75Delta** [comment]
For me I have to click this button to start the sync process, are you doing this?

![image](https://user-images.githubusercontent.com/33314913/186495084-3c62cc3f-1539-43d8-a502-45b4258abc43.png)

Though I've had to fork the driver and edit it a little before it worked on my device (Touch HD 2) so I might've accidentaly changed something.

**Comment by Engberg-Jacob** [comment]
Hm - I don't have a card in my device, but clicking the same button simply under "Device" doesn't help either, unfortunately

**Comment by zisoft** [comment]
> Though I've had to fork the driver and edit it a little before it worked on my device (Touch HD 2) so I might've accidentaly changed something.

@CZ75Delta  That's interesting, what did you need to change?

**Comment by CZ75Delta** [comment]
You can look at the diff in my fork, but a quick summary:
- COMPLETED_TS is not a column in my BOOKS_SETTINGS (My DB is explorer-2 if that makes any difference) 
- The way you look up the folder name didn't work for me as I'd get a path such as /mnt/ext2/Kazuki, Miya when I'd need a path from the Windows pov (i.e. I:/Kazuki, Miya) 
- I had to change the IDs for my device like so:
    - VENDOR_ID   = [0x1f85]
    - PRODUCT_ID  = [0x6a42]
    - BCD         = None
- BOOKS_FAST_HASHES is not a table in my database

I also disabled the pdf metadata corrections as I didn't know if they worked correctly and I wanted to go sleep :)

**Comment by zisoft** [comment]
Just released Version 1.0.3, please install and test.

If it still does not detect your device, please deactivate the other PocketBook device drivers in Calibre.

**Comment by Engberg-Jacob** [comment]
Thank you! Now the device is detected, and if a book is read on the Pocketbook, it is now marked as such in the userdefined column. 

But when I mark a book as read on Calibre, it's not imported and marked as such on the Pocketbook. Is this out-of-scope?

But still, thanks for your hard work

**Comment by zisoft** [comment]
It is a two-way sync, no matter where you mark a book as read, the status ist synchronized both on the device and in Calibre.
Of course, you need to perform a sync when you have marked a book as read in Calibre. This is done automatically when you connect the device the next time, or manually as described above.

Please let me know if it it works so I can close the issue.

**Comment by Engberg-Jacob** [comment]
Hmm - it doesn't seem to update now. I connect the device, but no matter where the read-status is changed, it doesn't update on the other one. Eg. if I mark something as read on the device, now it doesn't import that to Calibre, and when something is marked as read on Calibre, it isn't on the device, despite updating cached metadata.

Which is weird, because earlier, when it first detected the device, Calibre imported the read-status from the device, and it doesn't now. I will test a bit more, and return, but I am not sure why it was working, and now isn't, as the action I took to get it to import from the device hasn't changed.

**Comment by Engberg-Jacob** [comment]
I've tested it a bit more, and neither connecting the device, or using the "Update cached metadata on device" updates the read status either place. Since it is supposed to be automatic, and I can't recreate the events that made it happen the first time, I am at a lost what to do. But thanks for you help so far.

**Comment by zisoft** [comment]
Are you sure you have actually uninstalled any previous version of my plugin?

**Comment by Engberg-Jacob** [comment]
Sorry for the late response, yes. The only version installed is the 1.0.3 version. Does the plugin manage EPUB2 different vs. EPUB3?

Because it reacted only tried to change some of the version, and that didn't change anything. Is there any diagonstic tool that I can use to give more information?

**Comment by overdetermined** [comment]
So I am also having a Problem with the Touch Lux 5, and this the output in debug mode:

```Found device nodes: /dev/sdb None None
DEBUG:    0.0 PB_IMPROVED: open()
DEBUG:    0.0 PB_IMPROVED: read lookup name: #read
DEBUG:    0.0 PB_IMPROVED: Found database at path /run/media/vls/PB628/system/explorer-3/explorer-3.db
DEBUG:    0.0 PB_IMPROVED: database cleanup
DEBUG:    0.1 PB_IMPROVED: 21 rows from books_settings deleted
DEBUG:    0.1 PB_IMPROVED: 96 rows from booktogenre deleted
DEBUG:    0.1 PB_IMPROVED: 0 rows from social deleted
DEBUG:    0.1 PB_IMPROVED: 98 rows from books_impl deleted
DEBUG:    0.1 PB_IMPROVED: 1 rows from books_fast_hashes deleted
Job: 0 Get device information started
DeviceJob: 0 Get device information done, calling callback
DeviceJob: 0 Get device information callback returned
Job: 1 Set library information started
DeviceJob: 1 Set library information done, calling callback
DeviceJob: 1 Set library information callback returned
Job: 2 Get list of books on device started
DEBUG:    1.0 USBMS: Fetching list of books from device. Device= POCKETBOOK_IMPROVED oncard= None
DEBUG:    1.0 USBMS: dirs are: /run/media/vls/PB628/ Books
Job: 0 Get device information finished
No details available.
Job: 1 Set library information finished
No details available.```

**Comment by zisoft** [comment]
Did you actually deactivate the other pocketbook device drivers in calibre?

**Comment by overdetermined** [comment]
A valid question. Yes, I uninstalled all other plugins. However, today I wanted to verify this problem again and turns out a different usb port does not seem to have any problems. This other USB port happens to be USB 3.0 (while the one I was originally testing with USB 2. Not sure how this would be related but it could a problem that is unique to my computer or to the PocketBook. IFF I have some more time I may investigate further. Thanks for the plugin and the support. 

One thing that now comes up is that I am not able to see highlights I made in my PocketBook in Calibre... I will look into this next.
